File: LLD/MusicStreamingSystem.java
package com.spotify.lld;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/*
 * ==============================================================================================
 * MUSIC STREAMING SERVICE (SPOTIFY) LOW LEVEL DESIGN
 * ==============================================================================================
 * 
 * Key Features:
 * 1. Catalog Management: Songs, Albums, Artists.
 * 2. Playback: Player functionality (Play, Pause, queue).
 * 3. User Interaction: Playlists, Following Artists.
 * 4. Recommendation: Basic strategy based on Genre.
 * 5. Search: Find content by name.
 * 
 * Design Patterns:
 * 1. Singleton: StreamingService (Facade).
 * 2. Strategy: RecommendationStrategy.
 * 3. State: PlayerState (Playing, Paused).
 * 4. Composite: (Ideally Album contains Songs, Playlist contains Songs).
 * 
 * Class Design Diagram:
 * ---------------------
 * [StreamingService] "1" *-- "1" [Catalog]
 * [StreamingService] "1" *-- "*" [User]
 * [User] "1" *-- "1" [Player]
 * [User] "1" *-- "*" [Playlist]
 * [Catalog] "1" *-- "*" [Artist]
 * [Artist] "1" *-- "*" [Album]
 * [Album] "1" *-- "*" [Song]
 * [Player] ..> [Song]
 *
 * Class Details:
 * ---------------------
 * 1. StreamingService
 *    - Role: Facade.
 *    - Methods: search(), recommend().
 *
 * 2. Player
 *    - Role: Controls audio playback.
 *    - Attributes: currentSong, queue, state.
 *
 * 3. Catalog
 *    - Role: Index of all music.
 * 
 * 4. Song
 *    - Attributes: id, title, duration, genre.
 */

public class MusicStreamingSystem {
    public static void main(String[] args) {
        System.out.println("--- Spotify System Demo ---");
        
        StreamingService spotify = StreamingService.getInstance();

        // 1. Setup Catalog
        Artist a1 = new Artist("A1", "The Weeknd");
        Song s1 = new Song("S1", "Blinding Lights", "Pop", 200);
        Song s2 = new Song("S2", "Starboy", "Pop", 220);
        Album alb1 = new Album("AL1", "After Hours");
        
        alb1.addSong(s1); alb1.addSong(s2);
        a1.addAlbum(alb1);
        
        spotify.catalog.addArtist(a1); // Simplified adding

        // 2. User & Playlist
        User u1 = new User("U1", "Saurabh");
        spotify.registerUser(u1);
        
        Playlist p1 = u1.createPlaylist("Favorites");
        p1.addSong(s1);
        System.out.println("Playlist Created: " + p1.getName() + " Size: " + p1.getSongs().size());

        // 3. Playback
        System.out.println("\n[Action] User starts playing 'Blinding Lights'");
        u1.getPlayer().play(s1);
        u1.getPlayer().pause();

        // 4. Recommendation
        System.out.println("\n[Action] Recommendation for User based on 'Pop' history");
        // Simulate history
        u1.addToHistory(s1);
        List<Song> recs = spotify.getRecommendations(u1);
        System.out.println("Recommended Songs: " + recs.stream().map(Song::getTitle).collect(Collectors.toList()));
    }
}

// ==========================================
// Strategies
// ==========================================

interface RecommendationStrategy {
    List<Song> recommend(User user, Catalog catalog);
}

class GenreRecommendation implements RecommendationStrategy {
    @Override
    public List<Song> recommend(User user, Catalog catalog) {
        // Find most frequent genre in history
        // Simplified: Just pick genre of last played song
        if(user.getHistory().isEmpty()) return Collections.emptyList();
        
        Song lastPlayed = user.getHistory().get(user.getHistory().size()-1);
        String genre = lastPlayed.getGenre();
        
        return catalog.searchSongsByGenre(genre).stream()
                .filter(s -> !s.getId().equals(lastPlayed.getId())) // Don't recommend same song
                .collect(Collectors.toList());
    }
}

// ==========================================
// Domain Models
// ==========================================

class Song {
    String id, title, genre;
    int durationSec;
    public Song(String id, String t, String g, int d) {
        this.id = id; this.title = t; this.genre = g; this.durationSec = d;
    }
    public String getTitle() { return title; }
    public String getId() { return id; }
    public String getGenre() { return genre; }
}

class Album {
    String id, title;
    List<Song> songs;
    public Album(String id, String t) { 
        this.id = id; this.title = t; 
        this.songs = new ArrayList<>();
    }
    public void addSong(Song s) { songs.add(s); }
    public List<Song> getSongs() { return songs; }
}

class Artist {
    String id, name;
    List<Album> albums;
    public Artist(String id, String n) {
        this.id = id; this.name = n;
        this.albums = new ArrayList<>();
    }
    public void addAlbum(Album a) { albums.add(a); }
    public String getName() { return name; }
    public List<Album> getAlbums() { return albums; }
}

class Playlist {
    String name;
    List<Song> songs;
    public Playlist(String n) { this.name = n; this.songs = new ArrayList<>(); }
    public void addSong(Song s) { songs.add(s); }
    public List<Song> getSongs() { return songs; }
    public String getName() { return name; }
}

class Player {
    Song currentSong;
    boolean isPlaying;
    int currentSeekPosition;
    
    public void play(Song s) {
        this.currentSong = s;
        this.isPlaying = true;
        System.out.println("Playing: " + s.getTitle());
    }
    public void pause() {
        this.isPlaying = false;
        System.out.println("Paused: " + (currentSong!=null ? currentSong.getTitle() : "None"));
    }
    public void seek(int seconds) {
        this.currentSeekPosition = seconds;
    }
}

class User {
    String id, name;
    Player player;
    List<Playlist> playlists;
    List<Song> history;

    public User(String id, String name) {
        this.id = id; this.name = name;
        this.player = new Player();
        this.playlists = new ArrayList<>();
        this.history = new ArrayList<>();
    }
    
    public Playlist createPlaylist(String name) {
        Playlist p = new Playlist(name);
        playlists.add(p);
        return p;
    }
    
    public void addToHistory(Song s) { history.add(s); }
    public List<Song> getHistory() { return history; }
    public Player getPlayer() { return player; }
}

// ==========================================
// Catalog & System
// ==========================================

class Catalog {
    Map<String, Artist> artists = new ConcurrentHashMap<>();
    Map<String, Song> songs = new ConcurrentHashMap<>();
    
    public void addArtist(Artist a) {
        artists.put(a.id, a);
        for(Album alb : a.getAlbums()) {
            for(Song s : alb.getSongs()) {
                songs.put(s.id, s);
            }
        }
    }
    
    public List<Song> searchSongsByGenre(String genre) {
        return songs.values().stream()
                .filter(s -> s.getGenre().equalsIgnoreCase(genre))
                .collect(Collectors.toList());
    }
}

class StreamingService {
    private static StreamingService instance;
    public Catalog catalog;
    private Map<String, User> users;
    private RecommendationStrategy recStrategy;

    private StreamingService() {
        catalog = new Catalog();
        users = new HashMap<>();
        recStrategy = new GenreRecommendation();
    }

    public static synchronized StreamingService getInstance() {
        if(instance == null) instance = new StreamingService();
        return instance;
    }

    public void registerUser(User u) { users.put(u.id, u); }
    
    public List<Song> getRecommendations(User user) {
        return recStrategy.recommend(user, catalog);
    }
}

File: LLD/MusicStreamingSystem.py
# fmt: off
# ==============================================================================
#  MUSIC STREAMING SYSTEM — ASCII CLASS DIAGRAM
# ==============================================================================
#
#  ┌──────────────────────────────────────────────────────────────────────────┐
#  │                    MUSIC STREAMING SYSTEM (Spotify-like)                 │
#  └──────────────────────────────────────────────────────────────────────────┘
#
#  ┌─────────────────────────┐    ┌─────────────────────────────────────┐
#  │    StreamingService     │    │             MusicCatalog            │
#  │       (Facade)          │    ├─────────────────────────────────────┤
#  ├─────────────────────────┤    │ + artists: Dict[str, Artist]        │
#  │ + catalog: MusicCatalog │───>│ + songs: Dict[str, Song]            │
#  │ + users: Dict           │    ├─────────────────────────────────────┤
#  ├─────────────────────────┤    │ + add_artist()                      │
#  │ + register_user()       │    │ + search(query): List[Song]         │
#  │ + play_song()           │    └──────────────────────────────────────┘
#  │ + get_recommendations() │
#  └─────────────────────────┘    ┌─────────────────────────────────────┐
#                                 │               Artist                │
#  ┌─────────────────────┐        ├─────────────────────────────────────┤
#  │       User          │        │ + id: str                           │
#  ├─────────────────────┤        │ + name: str                         │
#  │ + id: str           │        │ + songs: List[Song]                 │
#  │ + name: str         │        ├─────────────────────────────────────┤
#  │ + player: Player    │        │ + add_song()                        │
#  │ + playlists: Dict   │        └──────────────────────────────────────┘
#  └────────┬────────────┘
#           │ 1 owns              ┌─────────────────────────────────────┐
#           ▼                    │                Song                 │
#  ┌─────────────────────┐       ├─────────────────────────────────────┤
#  │       Player        │       │ + id: str                           │
#  ├─────────────────────┤       │ + title: str                        │
#  │ + current_song      │       │ + artist: Artist                    │
#  │ + history: List[]   │       │ + duration_secs: int                │
#  ├─────────────────────┤       └──────────────────────────────────────┘
#  │ + play(song)        │
#  │ + pause() / stop()  │
#  └─────────────────────┘    ┌─────────────────────────────────────┐
#                             │              Playlist               │
#  ┌───────────────────────┐  ├─────────────────────────────────────┤
#  │ RecommendationStrategy│  │ + id: str                           │
#  │    (ABC/Interface)    │  │ + name: str                         │
#  ├───────────────────────┤  │ + songs: List[Song]                 │
#  │ + recommend(user,     │  ├─────────────────────────────────────┤
#  │   catalog): List[Song]│  │ + add_song()                        │
#  └──────────┬────────────┘  │ + remove_song()                     │
#             │               └──────────────────────────────────────┘
#      ┌──────┴───────┐
#      │              │
#      ▼              ▼
#  ┌──────────────┐  ┌─────────────────────────────────┐
#  │HistoryBased  │  │    GenreBased (extendable)      │
#  │Recommendation│  │    Recommendation               │
#  ├──────────────┤  └─────────────────────────────────┘
#  │ returns songs│
#  │ from unheard │
#  │ artists      │
#  └──────────────┘
#
#  RELATIONSHIPS:
#  StreamingService ──1──> MusicCatalog          (central song/artist registry)
#  StreamingService ──*──> User                  (manages registered users)
#  User ──1──> Player                            (owns a media player)
#  User ──*──> Playlist                          (creates playlists)
#  MusicCatalog ──*──> Artist                    (indexes artists)
#  Artist ──*──> Song                            (owns songs)
#  Playlist ──*──> Song                          (references songs)
#  Player.history ──*──> Song                    (play history)
#  StreamingService uses RecommendationStrategy  (Strategy Pattern)
#  HistoryBasedRecommendation ──▷── RecommendationStrategy (implements)
# ==============================================================================
# fmt: on
import threading
from abc import ABC, abstractmethod
from enum import Enum
from typing import List, Dict, Optional

"""
==============================================================================================
MUSIC STREAMING SERVICE (SPOTIFY) LOW LEVEL DESIGN (INTERVIEW OPTIMIZED)
==============================================================================================

Key Features:
1. Catalog Management: Songs, Albums, Artists.
2. Playback: Player functionality (Play, Pause, Stop).
3. User Interaction: Playlists, Listening History.
4. Recommendation: Strategy-based recommendations (e.g., Genre-based).
5. Search: Find songs by name.

Design Patterns:
1. Facade: StreamingService (Central Controller).
2. Strategy: RecommendationStrategy.
3. State: Handled via PlayerState enum.

Class Design Diagram:
---------------------
[StreamingService] "1" *-- "1" [Catalog]
[StreamingService] "1" *-- "*" [User]
[User] "1" *-- "1" [Player]
[User] "1" *-- "*" [Playlist]
[Catalog] "1" *-- "*" [Artist]
[Artist] "1" *-- "*" [Album]
[Album] "1" *-- "*" [Song]
[Player] ..> [Song]

Class Details:
---------------------
1. StreamingService (Facade)
   - Role: Facade.
   - Methods: registerUser(), search(), getRecommendations().

2. Player
   - Role: Controls audio playback.
   - Attributes: currentSong, state (PlayerState).
   - Methods: play(), pause(), stop().

3. Catalog
   - Role: Index of all music.
   - Methods: addArtist(), searchSongs().

4. Song
   - Attributes: id, title, genre, duration.
"""

# ==========================================
# Enums
# ==========================================

class PlayerState(Enum):
    PLAYING = "PLAYING"
    PAUSED = "PAUSED"
    STOPPED = "STOPPED"

# ==========================================
# Domain Models
# ==========================================

class Song:
    """Represents a musical track."""
    def __init__(self, song_id: str, title: str, genre: str, duration_sec: int):
        self.id = song_id
        self.title = title
        self.genre = genre
        self.duration_sec = duration_sec

    def __repr__(self):
        return f"Song({self.title})"

class Album:
    def __init__(self, album_id: str, title: str):
        self.id = album_id
        self.title = title
        self.songs: List[Song] = []

    def add_song(self, song: Song):
        self.songs.append(song)

class Artist:
    def __init__(self, artist_id: str, name: str):
        self.id = artist_id
        self.name = name
        self.albums: List[Album] = []

    def add_album(self, album: Album):
        self.albums.append(album)

class Playlist:
    """User-created collection of songs."""
    def __init__(self, name: str):
        self.name = name
        self.songs: List[Song] = []

    def add_song(self, song: Song):
        self.songs.append(song)
        print(f"INFO: Added '{song.title}' to playlist '{self.name}'")

# ==========================================
# Player Component
# ==========================================

class Player:
    """Controls playback for a specific user."""
    def __init__(self, user_name: str):
        self.user_name = user_name
        self.current_song: Optional[Song] = None
        self.state = PlayerState.STOPPED

    def play(self, song: Song):
        self.current_song = song
        self.state = PlayerState.PLAYING
        print(f"INFO: [{self.user_name}] Playing: {song.title} ({song.duration_sec}s)")

    def pause(self):
        if self.state == PlayerState.PLAYING:
            self.state = PlayerState.PAUSED
            print(f"INFO: [{self.user_name}] Paused: {self.current_song.title}")

    def stop(self):
        self.state = PlayerState.STOPPED
        self.current_song = None
        print(f"INFO: [{self.user_name}] Stopped playback.")

# ==========================================
# User & Recommendation Logic
# ==========================================

class User:
    """Streaming service user."""
    def __init__(self, user_id: str, name: str):
        self.id = user_id
        self.name = name
        self.player = Player(name)
        self.playlists: Dict[str, Playlist] = {}
        self.history: List[Song] = []

    def create_playlist(self, name: str) -> Playlist:
        p = Playlist(name)
        self.playlists[name] = p
        return p

    def add_to_history(self, song: Song):
        self.history.append(song)

class RecommendationStrategy(ABC):
    @abstractmethod
    def recommend(self, user: User, catalog: 'Catalog') -> List[Song]:
        pass

class GenreRecommendation(RecommendationStrategy):
    """Recommends songs based on the user's most recently played genre."""
    def recommend(self, user: User, catalog: 'Catalog') -> List[Song]:
        if not user.history:
            return []
        genre = user.history[-1].genre
        history_ids = {s.id for s in user.history}
        return [s for s in catalog.songs.values()
                if s.genre.lower() == genre.lower() and s.id not in history_ids]

# ==========================================
# Catalog & Service
# ==========================================

class Catalog:
    """Index of all music content."""
    def __init__(self):
        self.artists: Dict[str, Artist] = {}
        self.songs: Dict[str, Song] = {}
        self._lock = threading.Lock()

    def add_artist(self, artist: Artist):
        with self._lock:
            self.artists[artist.id] = artist
            for album in artist.albums:
                for song in album.songs:
                    self.songs[song.id] = song
            print(f"INFO: Catalog updated with artist {artist.name}")

    def search_songs(self, query: str) -> List[Song]:
        return [s for s in self.songs.values() if query.lower() in s.title.lower()]

class StreamingService:
    """Facade for the Music Streaming System."""
    def __init__(self):
        self.catalog = Catalog()
        self.users: Dict[str, User] = {}
        self.rec_strategy: RecommendationStrategy = GenreRecommendation()
        print("INFO: StreamingService initialized.")

    def register_user(self, user: User):
        self.users[user.id] = user
        print(f"INFO: User {user.name} registered.")

    def get_recommendations(self, user: User) -> List[Song]:
        print(f"INFO: Generating recommendations for {user.name}...")
        return self.rec_strategy.recommend(user, self.catalog)

# ==========================================
# Main Execution / Demo
# ==========================================

if __name__ == "__main__":
    print("--- Starting Music Streaming Demo ---")

    spotify = StreamingService()

    # 1. Setup content
    s1 = Song("S1", "Blinding Lights", "Pop", 200)
    s2 = Song("S2", "Save Your Tears", "Pop", 215)
    s3 = Song("S3", "Starboy", "Electronic", 230)
    s4 = Song("S4", "Get Lucky", "Pop", 240)

    alb1 = Album("AL1", "After Hours")
    alb1.add_song(s1)
    alb1.add_song(s2)
    artist1 = Artist("A1", "The Weeknd")
    artist1.add_album(alb1)

    alb2 = Album("AL2", "Random Access Memories")
    alb2.add_song(s4)
    artist2 = Artist("A2", "Daft Punk")
    artist2.add_album(alb2)

    spotify.catalog.add_artist(artist1)
    spotify.catalog.add_artist(artist2)

    # 2. User Actions
    user1 = User("U1", "Saurabh")
    spotify.register_user(user1)

    playlist = user1.create_playlist("Chill Vibes")
    playlist.add_song(s1)

    print("[Action] User plays 'Blinding Lights'")
    user1.player.play(s1)
    user1.add_to_history(s1)

    print("[Action] User searches for 'Lucky'")
    results = spotify.catalog.search_songs("Lucky")
    print(f"INFO: Search Results: {results}")

    # Recommendations: user1 played Pop, so expect more Pop (s2, s4)
    recs = spotify.get_recommendations(user1)
    print(f"INFO: Recommendations for {user1.name}: {recs}")
